### PR TITLE
WIP Add meaningful exception types

### PR DESCRIPTION
**Pull request by ismaelga**

### CODE REVIEW

**Comment by buddhamagnet** [comment]
Should you not also be referencing BadConfig (for example) by namespace i.e. Quartz::BadConfig. And as there are only three do they justify being in a separate file - they could just be delineated in quartz.rb as they are empty?

**Comment by ismaelga** [comment]
Well, I used the namespace because both ArgumentError and Exception are ruby builtin classes, I wanted to explicitly tell that I was using Quartz's versions of those classes. Bu yeah, I have nothing against using the namespace also on BadConfig.

About what file they should be defined at, I don't know. I feel good about having them at exceptions.rb, I like to have the exceptions at the same place so I can easily see what kind of exceptions I should expect/handle.

File: lib/quartz/go_process.rb [hunk 1]
```diff
@@ -46,7 +46,7 @@ def block_until_server_starts
     delay = 0.1 # seconds
 
     loop do
-      raise 'RPC server not starting' if retries > max_retries
+      raise Quartz::Exception, 'RPC server not starting' if retries > max_retries
```
Comment (DavidHuie) [inline lib/quartz/go_process.rb]: Could we make this more specific? Perhaps `GoServerException`?

File: lib/quartz/go_process.rb [hunk 2]
```diff
@@ -66,7 +66,7 @@ def get_metadata
     response = read
 
     if response['error']
-      raise "Metadata error: #{read['error']}"
+      raise Quartz::Exception, "Metadata error: #{read['error']}"
```
Comment (DavidHuie) [inline lib/quartz/go_process.rb]: Something like `GoResponseError` would be fitting here.

File: lib/quartz/exceptions.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module Quartz
+  class Exception < StandardError
+  end
+
+  class BadConfig < Exception
```
Comment (DavidHuie) [inline lib/quartz/exceptions.rb]: I would prefer the name `ConfigError`. The message can then provide details as to why it's bad.

**Comment by DavidHuie** [comment]
bump @ismaelga

**Comment by ismaelga** [comment]
Hey!
Yeah, will do the changes you proposed. Sorry for the delay :)

**Comment by ismaelga** [comment]
rebased

**Pull request merged**